internal/messagequeue/donthavetimeoutmgr.go

```diff
@@ -0,0 +1,297 @@
+		timeout:                    defaultTimeout,
+		defaultTimeout:             defaultTimeout,
+		latencyMultiplier:          latencyMultiplier,
+		maxExpectedWantProcessTime: maxExpectedWantProcessTime,
+		onDontHaveTimeout:          onDontHaveTimeout,
+	}
+
+	return mqp
+}
+
+// Shutdown the dontHaveTimeoutMgr. Any subsequent call to Start() will be ignored
+func (dhtm *dontHaveTimeoutMgr) Shutdown() {
+	dhtm.shutdown()
+}
+
+// onShutdown is called when the dontHaveTimeoutMgr shuts down
+func (dhtm *dontHaveTimeoutMgr) onShutdown() {
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	// Clear any pending check for timeouts
+	if dhtm.checkForTimeoutsTimer != nil {
+		dhtm.checkForTimeoutsTimer.Stop()
+	}
+}
+
+// onStarted is called when the dontHaveTimeoutMgr starts.
+// It monitors for the context being cancelled.
+func (dhtm *dontHaveTimeoutMgr) onStarted() {
+	<-dhtm.ctx.Done()
+	dhtm.onShutdown()
+}
+
+// Start the dontHaveTimeoutMgr. This method is idempotent
+func (dhtm *dontHaveTimeoutMgr) Start() {
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	// Make sure the dont have timeout manager hasn't already been started
+	if dhtm.started {
+		return
+	}
+	dhtm.started = true
+
+	go dhtm.onStarted()
+
+	// If we already have a measure of latency to the peer, use it to
+	// calculate a reasonable timeout
+	latency := dhtm.peerConn.Latency()
+	if latency.Nanoseconds() > 0 {
+		dhtm.timeout = dhtm.calculateTimeoutFromLatency(latency)
+		return
+	}
+
+	// Otherwise measure latency by pinging the peer
+	go dhtm.measureLatency()
+}
+
+// measureLatency measures the latency to the peer by pinging it
+func (dhtm *dontHaveTimeoutMgr) measureLatency() {
+	// Wait up to defaultTimeout for a response to the ping
+	ctx, cancel := context.WithTimeout(dhtm.ctx, dhtm.defaultTimeout)
+	defer cancel()
+
+	// Ping the peer
+	res := dhtm.peerConn.Ping(ctx)
+	if res.Error != nil {
+		// If there was an error, we'll just leave the timeout as
+		// defaultTimeout
+		return
+	}
+
+	// Get the average latency to the peer
+	latency := dhtm.peerConn.Latency()
+
+	// Calculate a reasonable timeout based on latency
+	dhtm.lk.Lock()
+	dhtm.timeout = dhtm.calculateTimeoutFromLatency(latency)
+	dhtm.lk.Unlock()
+
+	// Check if after changing the timeout there are any pending wants that are
+	// now over the timeout
+	dhtm.checkForTimeouts()
+}
+
+// checkForTimeouts checks pending wants to see if any are over the timeout
+func (dhtm *dontHaveTimeoutMgr) checkForTimeouts() {
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	if len(dhtm.wantQueue) == 0 {
+		return
+	}
+
+	// Figure out which of the blocks that were wanted were not received
+	// within the timeout
+	more := true
+	expired := make([]cid.Cid, 0, len(dhtm.pending))
+	removeCount := 0
+	for i := 0; i < len(dhtm.wantQueue) && more; i++ {
+		c := dhtm.wantQueue[i]
+
+		// If the want has been cancelled, its cid will have been removed from
+		// the pending map
+		if started, ok := dhtm.pending[c]; ok {
```

Comment:
What if we cancel then re-add it? I think we might actually need a cancel flag that lives in the queue itself. That is:

```go
type myjob struct {
    c cid.Cid
    canceled bool
}

...

wantQueue []*myjob
pending map[cid.Cid]*myjob
```